flink-runtime/src/test/java/org/apache/flink/runtime/state/SharedStateRegistryTest.java

```diff
@@ -182,20 +194,113 @@ public void testRegisterChangelogStateBackendHandles() throws InterruptedExcepti
     }
 
     @Test
-    public void testUnregisterUnusedState() {
+    public void testUnregisterUnusedSavepointState() {
         SharedStateRegistry sharedStateRegistry = new SharedStateRegistryImpl();
         TestingStreamStateHandle handle = new TestingStreamStateHandle();
-        sharedStateRegistry.registerReference(new SharedStateRegistryKey("first"), handle, 1L);
-        sharedStateRegistry.registerReference(new SharedStateRegistryKey("first"), handle, 2L);
-        sharedStateRegistry.registerReference(new SharedStateRegistryKey("first"), handle, 3L);
+
+        registerInitialCheckpoint(
+                sharedStateRegistry,
+                RESTORED_STATE_ID,
+                CheckpointProperties.forSavepoint(false, SavepointFormatType.NATIVE));
+
+        sharedStateRegistry.registerReference(
```

Comment:
The test checks the set of checkpoints that are not used anymore. 
This set doesn't depend on the restore mode.

Restore mode was added to `registerAllAfterRestored` earlier; it affects whether the unused state entry will be discarded or simply removed from the in-memory collection.

Reply:
I see. I just think about this case.
The checkpoint 2 and 3 should be placeholder in the real world if no failure occurs so that the restored state will not be discarded in NO_CLAIM mode, right ?
In your case, it just be overided by new reference so that it will work.

Reply:
I'm not sure I fully understand your comment, but shared state is discarded either when it's not used, or a **different** entry is submitted for the same key (that's probably what you meant). 
There is no contract to use placeholder; state backend might also send the same state handle.
If a different non-placeholder entry is submitted then it depends on whether the existing entry is already a part of a completed checkpoint. If it is, then the new entry will be discarded; otherwise, the existing entry will be discarded.